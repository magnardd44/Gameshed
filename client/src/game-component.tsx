import * as React from 'react';
import { Component } from 'react-simplified';
import {
  Alert,
  Card,
  Row,
  Column,
  Form,
  Button,
  ThumbNail,
  Container,
  ColumnCentre,
  Linebreak,
} from './widgets';
import { NavLink } from 'react-router-dom';
import { gameService, reviewService, Game, Review } from './services';
import { createHashHistory } from 'history';

const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully saving a student

export class GameCard extends Component {
  game: Game = {
    game_id: 0,
    game_title: '',
    genre: [],
    genre_id: 0,
    platform: [],
    game_description: '',
  };
  description =
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco...';
  render() {
    return (
      <Card title={this.game.game_title}>
        <h6 className="card-subtitle mb-2 text-muted">
          Terningkast:
          <ThumbNail small img="https://cdn-icons-png.flaticon.com/512/220/220725.png"></ThumbNail>
        </h6>
        <Row>
          <Column width={2}>
            <ThumbNail img="https://cdn-icons-png.flaticon.com/512/686/686589.png"></ThumbNail>
          </Column>
          <Column width={6}>
            {this.game.game_description}
            <Linebreak></Linebreak>
          </Column>
          <Column width={2}></Column>
        </Row>
        <Linebreak></Linebreak>
        <Row>
          <Column>Sjanger: {this.game.genre}</Column>
        </Row>
        <Linebreak></Linebreak>
        <Row>
          <Column>Platform: {this.game.platform}</Column>
        </Row>
      </Card>
    );
  }
  /** return axios
   */
  mounted() {
    gameService.getAll().then((result) => {
      this.game = result[0];
      console.log(this.game);
    });
  }
}
