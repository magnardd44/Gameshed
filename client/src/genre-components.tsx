import * as React from 'react';
import { Component, sharedComponentData } from 'react-simplified';
import {
  Alert,
  Card,
  Row,
  Column,
  Form,
  Button,
  ThumbNail,
  Container,
  ColumnCentre,
  Linebreak,
  FormContainer,
  FormGroup,
  CategoryCard,
} from './widgets';
import { NavLink } from 'react-router-dom';
import { gameService, Game } from './services/game-services';
import { Genre } from './services/genre-service';
import { createHashHistory } from 'history';
import axios from 'axios';
import { genreService } from './services/genre-service';
import { Review, reviewService } from './services/review-service';

const history = createHashHistory();

export class Category extends Component {
  genres: Genre[] = [];
  reviews: Review[] = [];
  games: Game[] = [];
  game: Game = {
    game_id: 0,
    game_title: '',
    genre: 0,
    genre_id: 0,
    platform: 0,
    game_description: '',
    igdb_id: 0,
    genres: [],
    platforms: [],
  };
  review: Review = {
    review_id: 0,
    game_id: 0,
    game_title: '',
    review_title: '',
    text: '',
    user_id: 0,
    rating: 0,
    published: false,
    genre_id: 0,
    relevant: 0,
    platform_id: 0,
  };
  state = { isHidden: false };
  genreCall(id: number) {
    reviewService
      .getGenre(id)
      .then((data) => {
        console.log(data);
        this.reviews = data;
      })
      .catch((error) => Alert.danger('Error retrieving reviews: ' + error.message));
    this.setState({ isHidden: false });
  }
  render() {
    return (
      <Container>
        Velg kategori/sjanger:
        <Row>
          <ColumnCentre width={12} smwidth={4} mdwidth={2}>
            {this.genres.map((genre) => (
              <CategoryCard
                key={genre.genre_id}
                img={
                  genre.genre_img
                    ? 'https://helenaagustsson.github.io/INFT2002-images/images/' +
                      genre.genre_img +
                      '.png'
                    : 'https://helenaagustsson.github.io/INFT2002-images/images/' +
                      genre.genre_name +
                      '.png'
                }
              >
                <Button.Success onClick={() => this.genreCall(genre.genre_id)}>
                  {genre.genre_name}
                </Button.Success>
              </CategoryCard>
            ))}
          </ColumnCentre>
          <ColumnCentre>
            {this.state.isHidden ? null : (
              <div className="sticky-top">
                {this.reviews.map((review, index) => (
                  <Row key={index}>
                    <Column>{review.game_title}</Column>
                    <Column>
                      <NavLink to={'/publishedReviews/' + review.review_id}>
                        {review.review_title}
                      </NavLink>
                    </Column>
                    <Column>{review.rating}</Column>
                  </Row>
                ))}
              </div>
            )}
          </ColumnCentre>
        </Row>
      </Container>
    );
  }
  mounted() {
    genreService
      .getAll()
      .then((genres) => {
        this.genres = genres;
      })
      .catch((error) => Alert.danger('Error getting reviews: ' + error.message));
  }
}
