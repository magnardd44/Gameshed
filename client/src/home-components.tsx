import * as React from 'react';
import { Component, sharedComponentData } from 'react-simplified';
import {
  Alert,
  Card,
  Row,
  Column,
  Form,
  Button,
  ThumbNail,
  Container,
  ColumnCentre,
  Linebreak,
  FormContainer,
  FormGroup,
  CategoryCard,
} from './widgets';
import { NavLink } from 'react-router-dom';
import { Genre } from './services/genre-service';
import { createHashHistory } from 'history';
import axios from 'axios';
import { genreService } from './services/genre-service';
import userService from './services/user-service';

export const history = createHashHistory();

export class Home extends Component {
  render() {
    return (
      <Container textalign={'center'}>
        <Row>
          <ColumnCentre width={6}>
            <Card title={'Søk etter spill'}>
              <Button.Success
                onClick={() => {
                  history.push('/search');
                }}
              >
                Søk
              </Button.Success>
            </Card>
          </ColumnCentre>
          <ColumnCentre width={6}>
            <Card title={'Søk etter anmeldelser'}>
              <Button.Success
                onClick={() => {
                  history.push('/reviews');
                }}
              >
                Søk
              </Button.Success>
            </Card>
          </ColumnCentre>
        </Row>
        <Linebreak></Linebreak>
        <Row>
          <ColumnCentre width={userService.token ? 12 : 6}>
            <Card title={'Legg til nytt spill'}>
              <Button.Success
                onClick={() => {
                  history.push('/addGame');
                }}
              >
                Legg til spill
              </Button.Success>
            </Card>
          </ColumnCentre>
          {userService.token ? (
            ''
          ) : (
            <ColumnCentre width={6}>
              <Card title="Registrer ny konto">
                <Button.Success
                  onClick={() => {
                    history.push('/user');
                  }}
                >
                  Registrer deg
                </Button.Success>
              </Card>
            </ColumnCentre>
          )}
        </Row>
      </Container>
    );
  }
}

export class ReviewHome extends Component {
  render() {
    return (
      <Container textalign={'center'}>
        <Row>
          <ColumnCentre width={6}>
            <Card title={'Sorter etter sjanger'}>
              <Button.Success
                onClick={() => {
                  history.push('/reviews-by-genre');
                }}
              >
                Søk
              </Button.Success>
            </Card>
          </ColumnCentre>
          <ColumnCentre width={6}>
            <Card title={'Sorter etter plattform'}>
              <Button.Success
                onClick={() => {
                  history.push('/reviews-by-platform');
                }}
              >
                Søk
              </Button.Success>
            </Card>
          </ColumnCentre>
        </Row>
        <Linebreak></Linebreak>
        <Row>
          <ColumnCentre width={6}>
            <Card title={'Topp ti anmeldelser'}>
              <Button.Success
                onClick={() => {
                  history.push('/reviews-by-stars');
                }}
              >
                Søk
              </Button.Success>
            </Card>
          </ColumnCentre>
          <ColumnCentre width={6}>
            <Card title={'Siste ti anmeldelser?'}>
              <Button.Success
                onClick={() => {
                  history.push('/reviews-by-date');
                }}
              >
                Søk
              </Button.Success>
            </Card>
          </ColumnCentre>
        </Row>
        <Linebreak></Linebreak>
        <Row>
          <ColumnCentre width={12}>
            <Card title={'Alle anmeldelser'}>
              <Button.Success
                onClick={() => {
                  history.push('/publishedReviews');
                }}
              >
                Alle anmeldelser
              </Button.Success>
            </Card>
          </ColumnCentre>
        </Row>
      </Container>
    );
  }
}
