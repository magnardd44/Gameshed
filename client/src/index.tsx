import * as React from 'react';
import { Component } from 'react-simplified';
import ReactDOM from 'react-dom';
import {
  Card,
  Alert,
  Row,
  Form,
  Column,
  Button,
  NavBar,
  Header,
  Container,
  ColumnCentre,
  SearchResult,
} from './widgets';
import axios from 'axios';
import { HashRouter, Route } from 'react-router-dom';
import { AddReview, PublishReview } from './review-components';
import { Hash } from 'crypto';
import { createHashHistory } from 'history';

axios.defaults.baseURL = 'http://localhost:3000/api/v2';

const history = createHashHistory();

class NavHeader extends Component {
  render() {
    return (
      <div>
        <NavBar brand="GameShed">
          <NavBar.Link to="/games">Games</NavBar.Link>
        </NavBar>
        <Header></Header>
      </div>
    );
  }
}

class Search extends Component {
  input = '';
  render() {
    return (
      <>
        <Container textalign={'centre'}>
          <Row>
            <ColumnCentre width={6} offset={2}>
              <Form.Input
                type={this.input}
                value={this.input}
                placeholder="Søk etter et spill"
                onChange={(event) => (this.input = event.currentTarget.value)}
              />
            </ColumnCentre>
            <ColumnCentre width={2}>
              <Button.Success
                onClick={() => {
                  this.search();
                }}
              >
                Søk
              </Button.Success>
            </ColumnCentre>
          </Row>
        </Container>
      </>
    );
  }
  search() {
    console.log('search');
    history.push('/results');
  }
}

class SearchListings extends Component {
  render() {
    return (
      <Container>
        Søkeresultater:
        <SearchResult></SearchResult>
        <SearchResult></SearchResult>
        <SearchResult></SearchResult>
      </Container>
    );
  }
}

ReactDOM.render(
  <HashRouter>
    <div>
      <NavHeader />
      <Route exact path="/" component={Search}></Route>
      <Route exact path="/results" component={SearchListings}></Route>
      <Route exact path="/addReview" component={AddReview} />
      <Route exact path="/publishReview/:id" component={PublishReview} />
    </div>
  </HashRouter>,
  document.getElementById('root')
);

/*
class Start extends Component {
  input = '';
  stdout = '';
  stderr = '';
  errCode: number | null = null;
}



  render() {
    return (
      <>
        <Card title="GameShed">
          <Row>
            <Column>
              <Form.Label>Søk:</Form.Label>
              <Form.Textarea
                value={this.input}
                onChange={(event) => (this.input = event.currentTarget.value)}
              />
            </Column>
          </Row>
          <Button.Success
            onClick={() => {
              axios
                .post<{ exitStatus: number; stdout: string; stderr: string }>('/run', {
                  language: 'js',
                  source: this.input,
                })
                .then((response) => {
                  this.errCode = response.data.exitStatus;
                  this.stdout = response.data.stdout;
                  this.stderr = response.data.stderr;
                })
                .catch((error: Error) => Alert.danger('Could not run app.js: ' + error.message));
            }}
          >
            Søk etter spill
          </Button.Success>
        </Card>
        <Card title="Standard output">{this.stdout}</Card>
        <Card title="Standard error">{this.stderr}</Card>
        <Card title="Errorcode">{this.errCode}</Card>
      </>
    );
  }
}


const root = document.getElementById('root');
if (root)
  ReactDOM.render(
    <HashRouter>
      <Alert />

      <Route exact path="/" component={Start} />
      
    </HashRouter>,
    root
  );
 
 */
