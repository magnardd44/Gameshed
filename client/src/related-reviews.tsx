import * as React from 'react';
import { Component } from 'react-simplified';
import {
  Alert,
  Card,
  Row,
  Column,
  Form,
  Button,
  ThumbNail,
  Container,
  ColumnCentre,
  Linebreak,
  ReviewCard,
} from './widgets';
import { NavLink } from 'react-router-dom';
import { gameService, Game } from './services/game-services';
import { gameService2, Game2 } from './services/game-services';
import { Genre, genreService } from './services/genre-service';
import { Review, reviewService } from './services/review-service';
import { createHashHistory } from 'history';
import { platform } from 'os';
import { Platform, platformService } from './services/platform-service';
import axios from 'axios';

const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully saving a student

export class RelatedReviews extends Component<{ game_id?: number; genre_id?: number }> {
  render() {
    return (
      <Container>
        <Row>
          <h5>Relaterte anmeldelser</h5>
          <ColumnCentre width={4}>
            <ReviewCard
              title={reviewService.relatedReview.review_title}
              subtitle={reviewService.relatedReview.game_title}
              terningkast={reviewService.relatedReview.rating}
              relevanse={reviewService.relatedReview.likes}
              text={reviewService.relatedReview.text}
            ></ReviewCard>
          </ColumnCentre>
        </Row>
        <Row>
          {reviewService.relatedReviews.map((review) => {
            <ColumnCentre width={4}>
              <ReviewCard
                title={review.review_title}
                subtitle={review.game_title}
                terningkast={review.rating}
                relevanse={review.likes}
                text={review.text}
              ></ReviewCard>
            </ColumnCentre>;
          })}
        </Row>
      </Container>
    );
  }

  mounted() {
    reviewService.getGenre(1).then((reviews) => {
      reviewService.relatedReviews = reviews;
      console.log('reviews' + reviewService.relatedReviews[0].review_title);
    });

    reviewService
      .get(1)
      .then((review) => {
        reviewService.relatedReview = review;
        console.log('reviews' + reviewService.relatedReviews[0].review_title);
      })
      .catch((error) => Alert.danger('Error retrieving reviews: ' + error.message));
  }
}
