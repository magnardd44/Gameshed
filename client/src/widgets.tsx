import * as React from 'react';
import { ReactNode, ChangeEvent } from 'react';
import { Component } from 'react-simplified';
import { NavLink } from 'react-router-dom';

/**
 * boostrap header
 * 
                  
 */
export class Heading extends Component<{ header?: boolean }> {
  render() {
    return (
      <div
        className="container text-center jumbotron my-3"
        style={
          this.props.header
            ? {
                color: '#24272bff',
                fontFamily: 'Road Rage, cursive',
                fontSize: '',
              }
            : {}
        }
      >
        {this.props.children}
        <div className="jumbotron">
          <h1 className="display-1 text-underline">GameShed</h1>
          <h2 className="display-5">Nettstedet for spillanmeldelser</h2>
        </div>
      </div>
    );
  }
}

/**
 * container for pages
 */

export class Container extends Component<{ textalign?: string }> {
  render() {
    return (
      <div
        className={
          'container my-5 mx-auto text' + (this.props.textalign ? '-' + this.props.textalign : '')
        }
      >
        {this.props.children}
      </div>
    );
  }
}

export class DropDown extends Component {
  render() {
    return (
      <div className="dropdown">
        <button
          className="btn btn-secondary dropdown-toggle"
          type="button"
          id="dropdownMenuButton"
          data-toggle="dropdown"
          aria-haspopup="true"
          aria-expanded="false"
        >
          Dropdown button
        </button>
        <div className="dropdown-menu" aria-labelledby="dropdownMenuButton">
          {this.props.children}
        </div>
      </div>
    );
  }
}

export class FormContainer extends Component {
  render() {
    return <form>{this.props.children}</form>;
  }
}

export class FormGroup extends Component {
  render() {
    return <div className="form-group">{this.props.children}</div>;
  }
}

export class ThumbNail extends Component<{ small?: boolean; img: string }> {
  render() {
    return (
      <img
        className="img-thumbnail"
        src={this.props.img}
        style={
          this.props.small
            ? {
                height: '16px',
                display: 'inline',
                padding: '0px',
              }
            : {}
        }
      />
    );
  }
}

export class Linebreak extends Component {
  render() {
    return <br />;
  }
}

/**
 * Renders an information card using Bootstrap classes.
 *
 * Properties: title
 */
export class Card extends Component<{ title: ReactNode }> {
  render() {
    return (
      <div className="card">
        <div className="card-body">
          <h5 className="card-title">{this.props.title}</h5>
          <div className="card-text">{this.props.children}</div>
        </div>
      </div>
    );
  }
}

export class CategoryCard extends Component<{ title?: ReactNode; img: string }> {
  render() {
    return (
      <div className="card border-success my-2">
        <div className="card-body text-center">
          <img
            className="card-img-top d-block mx-auto w-25"
            src={this.props.img}
            alt="Card image cap"
          />
        </div>
        {this.props.children}
      </div>
    );
  }
}

export class ReviewCard extends Component<{
  title: ReactNode;
  subtitle?: string;
  text?: string;
  terningkast?: number;
  relevanse?: number;
}> {
  render() {
    return (
      <div className="card border-success my-2">
        <Row>
          <Column width={2} offset={10} right>
            <ThumbNail
              small
              img={'https://helenaagustsson.github.io/INFT2002-images/images/star.png'}
            ></ThumbNail>

            {' ' + this.props.relevanse}
          </Column>
        </Row>
        <div className="card-body pt-0">
          <Row>
            <ColumnCentre width={11}>
              <h5 className="card-title">
                {this.props.title + ' '}
                <ThumbNail
                  small
                  img={
                    'https://helenaagustsson.github.io/INFT2002-images/images/dice-' +
                    this.props.terningkast +
                    '.png'
                  }
                ></ThumbNail>
              </h5>
              <h6 className="card-subtitle mb-2 text-muted">Spill: {this.props.subtitle}</h6>
            </ColumnCentre>
          </Row>
          <Row>
            <ColumnCentre width={12}>
              <div className="card-text">
                {this.props.text?.substr(0, 175) + '... '}
                {this.props.children}
              </div>
            </ColumnCentre>
          </Row>
        </div>
      </div>
    );
  }
}

/**
 * Renders a row using Bootstrap classes.
 */
export class Row extends Component {
  render() {
    return <div className="row">{this.props.children}</div>;
  }
}

export class ColumnCentre extends Component<{
  width?: number;
  offset?: number;
  smwidth?: number;
  mdwidth?: number;
}> {
  render() {
    return (
      <div
        className={
          'col' +
          (this.props.width ? '-' + this.props.width : '') +
          (this.props.smwidth ? ' col-sm-' + this.props.smwidth : '') +
          (this.props.mdwidth ? ' col-md-' + this.props.mdwidth : '') +
          (this.props.offset ? ' offset-' + this.props.offset : '')
        }
      >
        {this.props.children}
      </div>
    );
  }
}

/**
 * Renders a column with specified width using Bootstrap classes.
 *
 * Properties: width, right
 */
export class Column extends Component<{ width?: number; right?: boolean; offset?: number }> {
  render() {
    return (
      <div
        className={
          'col-sm' +
          (this.props.width ? '-' + this.props.width : '') +
          (this.props.offset ? ' offset-' + this.props.offset : '')
        }
      >
        <div className={'float-' + (this.props.right ? 'end' : 'start')}>{this.props.children}</div>
      </div>
    );
  }
}

/**
 * Renders a success button using Bootstrap styles.
 *
 * Properties: small, onClick
 */
class ButtonSuccess extends Component<{ small?: boolean; onClick: () => void }> {
  render() {
    return (
      <button
        type="button"
        className="btn btn-success"
        style={
          this.props.small
            ? {
                padding: '5px 5px',
                fontSize: '16px',
                lineHeight: '0.7',
                display: 'inline',
              }
            : {}
        }
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    );
  }
}

/**
 * Renders a danger button using Bootstrap styles.
 *
 * Properties: small, onClick
 */
class ButtonDanger extends Component<{ small?: boolean; onClick: () => void }> {
  render() {
    return (
      <button
        type="button"
        className="btn btn-danger"
        style={
          this.props.small
            ? {
                padding: '5px 5px',
                fontSize: '16px',
                lineHeight: '0.7',
              }
            : {}
        }
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    );
  }
}

/**
 * Renders a light button using Bootstrap styles.
 *
 * Properties: small, onClick
 */
class ButtonLight extends Component<{ small?: boolean; onClick: () => void }> {
  render() {
    return (
      <button
        type="button"
        className="btn btn-light"
        style={
          this.props.small
            ? {
                padding: '5px 5px',
                fontSize: '16px',
                lineHeight: '0.7',
              }
            : {}
        }
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    );
  }
}

/**
 * Renders a button using Bootstrap styles.
 *
 * Properties: onClick
 */
export class Button {
  static Success = ButtonSuccess;
  static Danger = ButtonDanger;
  static Light = ButtonLight;
}

/**
 * Renders a NavBar link using Bootstrap styles.
 *
 * Properties: to
 */
export class NavBarLink extends Component<{ to: string }> {
  render() {
    return (
      <NavLink className="nav-link" activeClassName="active" to={this.props.to}>
        {this.props.children}
      </NavLink>
    );
  }
}

/**
 * Renders a NavBar using Bootstrap classes.
 *
 * Properties: brand
 */
export class JustifiedNavBar extends Component<{ brand?: ReactNode; justify?: string }> {
  static Link = NavBarLink;

  render() {
    return (
      <nav className="navbar navbar-expand-sm navbar-light bg-light">
        <div className={'container-fluid justify-content-' + this.props.justify}>
          <NavLink
            className="navbar-brand"
            activeClassName="active"
            exact
            to="/"
            style={
              this.props.brand
                ? {
                    color: '#24272bff',
                    fontFamily: 'Road Rage, cursive',
                    fontSize: '2rem',
                  }
                : {}
            }
          >
            {this.props.brand}
          </NavLink>
          <div className="navbar-nav">{this.props.children}</div>
        </div>
      </nav>
    );
  }
}

export class NavBar extends Component<{ brand: ReactNode; justify?: string }> {
  static Link = NavBarLink;

  render() {
    return (
      <nav className="navbar navbar-expand-sm navbar-light bg-light">
        <div className="container-fluid justify-content-start">
          <NavLink className="navbar-brand" activeClassName="active" exact to="/">
            {this.props.brand}
          </NavLink>
          <div className="navbar-nav">{this.props.children}</div>
        </div>
      </nav>
    );
  }
}

/**
 * Renders a form label using Bootstrap styles.
 */
class FormLabel extends Component {
  render() {
    return <label className="col-form-label mx-2">{this.props.children}</label>;
  }
}

/**
 * Renders a form input using Bootstrap styles.
 */
class FormInput extends Component<{
  type: string;
  value: string | number;
  onChange: (event: ChangeEvent<HTMLInputElement>) => void;
  [prop: string]: any;
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, width, height, pattern
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { type, value, onChange, ...rest } = this.props;
    return (
      <input
        {...rest}
        className="form-control"
        type={this.props.type}
        value={this.props.value}
        onChange={this.props.onChange}
      />
    );
  }
}

/**
 * Renders a form textarea using Bootstrap styles.
 */
class FormTextarea extends React.Component<{
  value: string | number;
  onChange: (event: ChangeEvent<HTMLTextAreaElement>) => void;
  [prop: string]: any;
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, rows, cols
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { value, onChange, ...rest } = this.props;
    return <textarea {...rest} className="form-control" value={value} onChange={onChange} />;
  }
}

/**
 * Renders a form checkbox using Bootstrap styles.
 */
class FormCheckbox extends Component<{
  checked: boolean;
  onChange: (event: ChangeEvent<HTMLInputElement>) => void;
  [prop: string]: any;
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, width, height, pattern
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { checked, onChange, ...rest } = this.props;
    return (
      <input
        {...rest}
        className="form-check-input"
        type="checkbox"
        checked={checked}
        onChange={onChange}
      />
    );
  }
}

/**
 * Renders a form select using Bootstrap styles.
 */
class FormSelect extends Component<{
  value: string | number;
  onChange: (event: ChangeEvent<HTMLSelectElement>) => void;
  [prop: string]: any;
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, size.
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { value, onChange, children, ...rest } = this.props;
    return (
      <select {...rest} className="custom-select" value={value} onChange={onChange}>
        {children}
      </select>
    );
  }
}

/**
 * Renders form components using Bootstrap styles.
 */
export class Form {
  static Label = FormLabel;
  static Input = FormInput;
  static Textarea = FormTextarea;
  static Checkbox = FormCheckbox;
  static Select = FormSelect;
}

/**
 * Renders alert messages using Bootstrap classes.
 *
 * Students: this slightly more complex component is not part of curriculum.
 */
export class Alert extends Component {
  alerts: { id: number; text: ReactNode; type: string }[] = [];
  nextId: number = 0;

  render() {
    return (
      <div>
        {this.alerts.map((alert, i) => (
          <div
            key={alert.id}
            className={'alert alert-dismissible alert-' + alert.type}
            role="alert"
          >
            {alert.text}
            <button
              type="button"
              className="btn-close btn-sm"
              onClick={() => this.alerts.splice(i, 1)}
            />
          </div>
        ))}
      </div>
    );
  }

  /**
   * Show success alert.
   */
  static success(text: ReactNode) {
    // To avoid 'Cannot update during an existing state transition' errors, run after current event through setTimeout
    setTimeout(() => {
      let instance = Alert.instance(); // Get rendered Alert component instance
      if (instance) instance.alerts.push({ id: instance.nextId++, text: text, type: 'success' });
    });
  }

  /**
   * Show info alert.
   */
  static info(text: ReactNode) {
    // To avoid 'Cannot update during an existing state transition' errors, run after current event through setTimeout
    setTimeout(() => {
      let instance = Alert.instance(); // Get rendered Alert component instance
      if (instance) instance.alerts.push({ id: instance.nextId++, text: text, type: 'info' });
    });
  }

  /**
   * Show warning alert.
   */
  static warning(text: ReactNode) {
    // To avoid 'Cannot update during an existing state transition' errors, run after current event through setTimeout
    setTimeout(() => {
      let instance = Alert.instance(); // Get rendered Alert component instance
      if (instance) instance.alerts.push({ id: instance.nextId++, text: text, type: 'warning' });
    });
  }

  /**
   * Show danger alert.
   */
  static danger(text: ReactNode) {
    // To avoid 'Cannot update during an existing state transition' errors, run after current event through setTimeout
    setTimeout(() => {
      let instance = Alert.instance(); // Get rendered Alert component instance
      if (instance) instance.alerts.push({ id: instance.nextId++, text: text, type: 'danger' });
    });
  }
}
