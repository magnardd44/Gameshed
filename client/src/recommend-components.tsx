import * as React from 'react';
import { Component, sharedComponentData } from 'react-simplified';
import {
  Alert,
  Card,
  Row,
  Column,
  Form,
  Button,
  ThumbNail,
  Container,
  ColumnCentre,
  Linebreak,
  FormContainer,
  FormGroup,
  CategoryCard,
  ReviewCard,
} from './widgets';
import { NavLink } from 'react-router-dom';
import { Genre } from './services/genre-service';
import { createHashHistory } from 'history';
import axios from 'axios';
import { genreService } from './services/genre-service';
import { Review, reviewService } from './services/review-service';
import { RelatedReviews } from './related-reviews';

export const history = createHashHistory();

export class TopTenStars extends Component {
  reviews: Review[] = [];
  render() {
    return (
      <>
        <Container>
          <h3 className="text-center">Her kan du se de best likte anmeldelsene</h3>
          <Linebreak></Linebreak>
          <Row>
            <ColumnCentre>
              <ReviewList reviews={this.reviews}></ReviewList>
            </ColumnCentre>
          </Row>
        </Container>
      </>
    );
  }
  mounted() {
    reviewService
      .getTopTen()
      .then((reviews) => {
        this.reviews = reviews;
      })
      .catch((error) => Alert.danger('Error getting top ten reviews: ' + error.message));
  }
}

export class LastTen extends Component {
  reviews: Review[] = [];
  render() {
    return (
      <>
        <Container>
          <h3 className="text-center">Her kan du se de siste publiserte anmeldelsene</h3>
          <Linebreak></Linebreak>
          <Row>
            <ColumnCentre>
              <ReviewList reviews={this.reviews}></ReviewList>
            </ColumnCentre>
          </Row>
        </Container>
      </>
    );
  }
  mounted() {
    reviewService
      .getLastTen()
      .then((reviews) => {
        this.reviews = reviews;
      })
      .catch((error) => Alert.danger('Error getting top ten reviews: ' + error.message));
  }
}

export class ReviewList extends Component<{ reviews: Review[] }> {
  render() {
    return (
      <div className="sticky-top">
        {this.props.reviews.length == 0 ? (
          <Row>
            <Column>Ingen resultater</Column>
          </Row>
        ) : (
          this.props.reviews.map((review, index) => (
            <Row key={index}>
              <ReviewCard
                title={review.review_title}
                subtitle={review.game_title}
                terningkast={review.rating}
                relevanse={review.likes}
                text={review.text}
              >
                <Button.Success
                  small
                  onClick={() => {
                    history.push('/publishedReviews/' + review.review_id);
                  }}
                >
                  Les mer
                </Button.Success>
              </ReviewCard>
            </Row>
          ))
        )}
      </div>
    );
  }
}
