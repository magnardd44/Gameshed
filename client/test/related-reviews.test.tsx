import * as React from 'react';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import { shallow } from 'enzyme';
import { history, RelatedReviews } from '../src/related-reviews';

import { ColumnCentre, Form, Card, Alert, Button, Container, Column } from '../src/widgets';

const mockAdapter = new MockAdapter(axios);

describe('RelatedReviews component', () => {
  test('RelatedReviews component default', () => {
    const wrapper = shallow(<RelatedReviews />);

    expect(wrapper).toMatchSnapshot();
  });
});
