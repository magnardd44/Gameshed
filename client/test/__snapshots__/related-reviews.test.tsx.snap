// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelatedReviews component RelatedReviews component default 1`] = `
<Container>
  <Row>
    <h5>
      Relaterte anmeldelser
    </h5>
    <ColumnCentre
      width={4}
    >
      <ReviewCard
        relevanse={0}
        subtitle=""
        terningkast={0}
        text=""
        title=""
      />
    </ColumnCentre>
  </Row>
  <Row />
</Container>
`;
